File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cnt = 0
        start = head
        while head:
            cnt += 1
            head = head.next
        mid = cnt // 2
        
        while mid:
            start = start.next
            mid -= 1
        
        return start

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode find_len = head;
        ListNode ans = head;

        int len = 0;
        while (find_len.next != null) {
            find_len = find_len.next;
            len++;
        }
        int mid = 0;
        if (len % 2 == 1) {
            mid = len / 2 + 1;
        }
        else {
            mid = len / 2;
        }

        int start = 0;
        while (start < mid) {
            ans = ans.next;
            start++;
        }
        return ans;
    }
}